ambry-tools/src/main/java/com.github.ambry.tools/admin/NetworkClientUtils.java

```diff
@@ -0,0 +1,194 @@
+/**
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.tools.admin;
+
+import com.github.ambry.network.NetworkClient;
+import com.github.ambry.network.NetworkClientErrorCode;
+import com.github.ambry.network.RequestInfo;
+import com.github.ambry.network.ResponseInfo;
+import com.github.ambry.protocol.DeleteRequest;
+import com.github.ambry.protocol.GetRequest;
+import com.github.ambry.protocol.PutRequest;
+import com.github.ambry.protocol.RequestOrResponse;
+import com.github.ambry.protocol.RequestOrResponseType;
+import com.github.ambry.router.Callback;
+import com.github.ambry.router.FutureResult;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Helper class to send requests and receive responses via {@link NetworkClient}.
+ * Each caller to {@link #poll(List)} will register a future (and may be a callback) for every request that needs to be
+ * sent to the network client. On receiving the response, the {@link NetworkClientUtils} sets the result/exception in
+ * the future and calls the callback if not {@code null}
+ */
+public class NetworkClientUtils {
+  private final NetworkClient networkClient;
+  private final int pollTimeMs;
+  private final Map<Integer, RequestMetadata> correlationIdToRequestMetadataMap;
+  private final Lock lock;
+  private final NetworkClientPoller _networkClientPoller;
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  public NetworkClientUtils(NetworkClient networkClient, int pollTimeMs, long intervalToPollNetworkClientInMs) {
+    this.networkClient = networkClient;
+    this.pollTimeMs = pollTimeMs;
+    this.correlationIdToRequestMetadataMap = new HashMap<>();
+    this.lock = new ReentrantLock();
+    this._networkClientPoller = new NetworkClientPoller(intervalToPollNetworkClientInMs);
+    new Thread(_networkClientPoller).start();
+  }
+
+  /**
+   * Polls the {@link NetworkClient} for the list of {@link RequestInfo} sent it
+   * @param requestMetadataList
+   */
+  void poll(List<RequestMetadata> requestMetadataList) {
+    try {
+      lock.lock();
```

Comment:
why don't you use a synchronized method?
first of all, what is this lock for? are you having multiple threads call this method? you only need a single thread to send concurrent requests via the network client.